File: week1/Design_Pattern_Principles/exercise1_singelton/Logger.java
public class Logger {

    // This is where we’ll keep the one and only Logger instance
    private static Logger instance;

    // Constructor is private, so no one can make a new Logger using 'new'
    private Logger() {
        System.out.println("Logger Initialized");
    }

    // This is the only way to get the Logger
    public static Logger getInstance() {
        // If it doesn't exist, create it
        if (instance == null) {
            instance = new Logger();
        }
        return instance;
    }

    // A simple method to simulate logging messages
    public void log(String message) {
        System.out.println("[LOG] " + message);
    }
}

File: week1_Engineering_Concepts/Design_Pattern_Principles/exercise1_singelton/code/Logger.cs
using System;

namespace SingletonExample
{
    public class Logger
    {
        // Private static instance of Logger
        private static Logger instance;

        // Private constructor to prevent external instantiation
        private Logger()
        {
            Console.WriteLine("Logger Initialized");
        }

        // Public method to get the single instance
        public static Logger GetInstance()
        {
            if (instance == null)
            {
                instance = new Logger();
            }
            return instance;
        }

        // Method to log a message
        public void Log(string message)
        {
            Console.WriteLine("[LOG] " + message);
        }
    }
}
